DOCS-1916: Add overwrite fragment mods (#2622)
viamrobotics/docs: docs/fleet/configure-a-fleet.md

You must be an [organization owner](/fleet/rbac/#permissions) in order to create fragments.

A fragment can define one, several, or all resources on a machine.

> You can add multiple fragments to a single machine, or you can add additional resources to a machine that has already been configured with a fragment.

Author: Great question, but outside the scope of this PR! I'll make a ticket to follow up.

## Modify the config of a machine that uses a fragment

When you modify a fragment, those changes are pushed to all machines that use that fragment.

> If you need to modify the config of just one machine that uses a fragment you have two options:

Author: I think singular sounds a little bit more formal than necessary and I think just one makes sense--I'd like to keep as-is if that's okay!

### Use `fragment_mods`

You can modify fragment fields in your machine's raw JSON config by using [update operators](https://www.mongodb.com/docs/manual/reference/operator/update/positional/#---update-).
Viam supports all update operators except for `$setOnInsert`, `$`, `$[]`, and `$[<identifier>]`.

1. Navigate to your machine's **Config** tab.
2. Switch to **Raw JSON** mode.
3. Add a top-level section called `"fragment_mods"` (alongside the other top-level sections like `"components"` and `"fragments"`):

{{< tabs >}}
{{% tab name="Template" %}}

```json {class="line-numbers linkable-line-numbers"}
  "fragment_mods": [
    {
      "fragment_id": "<YOUR FRAGMENT ID>",
      "mods": [
        {
          <INSERT YOUR MODS HERE>
        }
      ]
    }
  ],
```

{{% /tab %}}
{{% tab name="Full example" %}}
This example assumes the fragment with ID `abcd7ef8-fa88-1234-b9a1-123z987e55aa` contains a motor configured with `"name": "motor1"`.

```json {class="line-numbers linkable-line-numbers"}
{
  "components": [],
  "fragment_mods": [
    {
      "fragment_id": "abcd7ef8-fa88-1234-b9a1-123z987e55aa",
      "mods": [

>         {

Reviewer: looking back it this - I think when the command is the same, it shouldn't be repeated - just have it as:

```

"$set": {
"components.motor1.attributes.max_rpm": 1818,
"components.motor1.attributes.pins.a": 30,
"components.motor1.attributes.board": "local"
}

```

Unless you were able to get it to work this way . Going to test this right now though to double check.

Reviewer: Yeah I think they may have to be stacked if the command is the same within the same fragment_mods entry - there's a duplicate key error in the json and it discards the second.

https://github.com/viamrobotics/docs/assets/51691446/f2b0f49c-1c7b-4a30-b52d-820f1c5eb6e4

Author: Oh interesting! That looks a lot cleaner. I didn't try combining them--I only tried it repeated, which did work also!

Reviewer: oh interesting 🤔 I wonder why mine didn't that's odd

>           "$set": {
>             "components.motor1.attributes.max_rpm": 1818
>           }
>         },
>         {
>           "$set": {
>             "components.motor1.attributes.pins.a": 30
>           }
>         },
>         {
>           "$set": {
>             "components.motor1.attributes.board": "local"
>           }
>         },

Author: ```suggestion
          "$set": {
            "components.motor1.attributes.max_rpm": 1818,
            "components.motor1.attributes.pins.a": 30,
            "components.motor1.attributes.board": "local"
          }
        },
```

        {
          "$unset": {
            "components.motor1.attributes.pins.pwm": 0
          }
        }
      ]
    }
  ],
  "fragments": ["abcd7ef8-fa88-1234-b9a1-123z987e55aa"]
}
```

{{% /tab %}}
{{< /tabs >}}

4. Edit the `fragment_id` value to match the ID of the fragment you want to modify, for example `"12345678-1a2b-9b8a-abcd987654321"`.

> 5. Add any mods you'd like to apply.

Author: Hmm okay I'll do this:
```suggestion
5. Add any update operators you'd like to apply to the fragment to the `mods` section.
```

   Click to view each example:

   {{%expand "Change the name and attributes of a component" %}}
   This example uses [`$set`](https://www.mongodb.com/docs/manual/reference/operator/update/set/#mongodb-update-up.-set) to make the following changes to the attributes of a motor named `motor1`:

   - Sets the `max_rpm` to `1818`.
   - Changes the name of `motor1` to `my_motor`.
     Note that this does not affect the other mods; you still use `motor1` for them.
   - Sets the pin number for pin `a` to `30`.
   - Sets the name of the board associated with this motor to `local`.

   ```json {class="line-numbers linkable-line-numbers"}
   "fragment_mods": [
    {
      "fragment_id": "abcd7ef8-fa88-1234-b9a1-123z987e55aa",
      "mods": [
        {

>           "$set": {

Reviewer: it would have to happen here as well

>           "$set": {
>             "components.motor1.attributes.max_rpm": 1818
>           }
>         },
>         {
>           "$set": {
>             "components.motor1.name": "my_motor"
>           }
>         },
>         {
>           "$set": {
>             "components.motor1.attributes.pins.a": 30
>           }
>         },
>         {
>           "$set": {
>             "components.motor1.attributes.board": "local"
>           }
>         }

Author: ```suggestion
          "$set": {
            "components.motor1.attributes.max_rpm": 1818,
            "components.motor1.name": "my_motor",
            "components.motor1.attributes.pins.a": 30,
            "components.motor1.attributes.board": "local"
          }
        }
```

   ```json {class="line-numbers linkable-line-numbers"}
   "fragment_mods": [
     {
       "fragment_id": "abcd7ef8-fa88-1234-b9a1-123z987e55aa",
       "mods": [
         {
          "$unset": {
            "components.motor1.attributes.pins.pwm": 0
          }
         }
       ]
     }
   ],
   ```

   {{% /expand%}}
   {{%expand "Modify dependencies" %}}
   This example uses [`$set`](https://www.mongodb.com/docs/manual/reference/operator/update/set/#mongodb-update-up.-set) to assign a new list of dependencies to a component named `rover_base2`.

   ```json {class="line-numbers linkable-line-numbers"}
   "fragment_mods": [
     {
       "fragment_id": "abcd7ef8-fa88-1234-b9a1-123z987e55aa",
       "mods": [
         {
          "$set": {
            "components.rover_base2.attributes.depends_on": ["local", "motor1"]
          }
         }
       ]
     }
   ],
   ```

   {{% /expand%}}
   {{%expand "Change motor pins from A and B to PWM and DIR" %}}
   This example uses [`$rename`](https://www.mongodb.com/docs/manual/reference/operator/update/rename/) to make the following changes to the attributes of a motor named `motor1` in the fragment:

>    - Takes whatever the pin number for pin `a` was and assigns that value to the PWM pin.

Author: Ooh I really like the word "retrieves"--that adds clarity/specificity. Also I like your wording that removes "whatever" and "was"--it sounds more concise and professional. I'll incorporate that but I will keep the second half of the sentence the same--I do think it's easier to parse as-is.
```suggestion
   - Retrieves the pin number for pin `a` and assigns that value to the PWM pin.
```

     Deletes the `pins.a` field.

>    - Takes whatever the pin number for pin `b` was and assigns that value to the DIR pin.

Author: ```suggestion
   - Retrieves the pin number for pin `b` was and assigns that value to the DIR pin.
```

     Deletes the `pins.b` field.

>    _If you want to change the `name` of a component (for example, `motor1`), use `$set`, as shown in the change the name of a component example.

Author: I'll swap the order of these two sentences--see below

>    `$rename` is for changing the key, not the value._

Author: Yes--if I understand your question correctly, the example here shows exactly that--two things being renamed simultaneously. I tested and it worked.

Yeah I can add that clarification. I'll also swap the order of these two italicized sentences, which I think will help with your comment above this too.

>    _If you want to change the `name` of a component (for example, `motor1`), use `$set`, as shown in the change the name of a component example.
>    `$rename` is for changing the key, not the value._

Author: ```suggestion
   _`$rename` is for changing an attribute's key, not its value.
   If you want to change the `name` of a component (for example, `motor1`), use `$set`, as shown in the change the name of a component example._
```

   ```json {class="line-numbers linkable-line-numbers"}
   "fragment_mods": [
    {
      "fragment_id": "abcd7ef8-fa88-1234-b9a1-123z987e55aa",
      "mods": [
        {

>           "$rename": {

Reviewer: stacking would have to be here too. Sorry I missed this earlier.
Also forgot to mention that the docs look really nice - thank you 😊

>             "components.motor1.attributes.pins.a": "components.motor1.attributes.pins.pwm"
>           }
>         },
>         {
>           "$rename": {
>             "components.motor1.attributes.pins.b": "components.motor1.attributes.pins.dir"
>           }

Author: ```suggestion
            "components.motor1.attributes.pins.a": "components.motor1.attributes.pins.pwm",
            "components.motor1.attributes.pins.b": "components.motor1.attributes.pins.dir"
          }
```
